addons/subscription/models/subscription.py

```diff
@@ -0,0 +1,120 @@
+    note = fields.Text(string='Notes', help="Description or Summary of Subscription")
+
+    @api.model
+    def _auto_end(self):
+        super(Subscription, self)._auto_end()
+        # drop the FK from subscription to ir.cron, as it would cause deadlocks
+        # during cron job execution. When model_copy() tries to write() on the subscription,
+        # it has to wait for an ExclusiveLock on the cron job record, but the latter
+        # is locked by the cron system for the duration of the job!
+        # FIXME: the subscription module should be reviewed to simplify the scheduling process
+        #        and to use a unique cron job for all subscriptions, so that it never needs to
+        #        be updated during its execution.
+        self.env.cr.execute("ALTER TABLE %s DROP CONSTRAINT %s" % (self._table, '%s_cron_id_fkey' % self._table))
+
+    @api.multi
+    def set_process(self):
+        for subscription in self:
+            res = {'name': subscription.name, 'interval_number': subscription.interval_number, 'interval_type': subscription.interval_type, 'numbercall': subscription.exec_init, 'nextcall': subscription.date_init, 'model': 'subscription.subscription', 'args': repr([[subscription.id]]), 'function': 'model_copy', 'priority': 6, 'user_id': subscription.user_id.id}
+            cron = self.env['ir.cron'].create(res)
+            subscription.write({'cron_id': cron.id, 'state': 'running'})
+
+    @api.multi
+    def model_copy(self):
+        document_model = self.env['subscription.document']
+        history_model = self.env['subscription.subscription.history']
+        for subscription in self.filtered(lambda sub: sub.cron_id):
+            if not subscription.doc_source.exists():
+                raise UserError(_('Please provide another source document.\nThis one does not exist!'))
+            model_name = subscription.doc_source._name
+
+            docsource_data = {'state': 'draft'}
+
+            datefields = document_model.search([('model_id.model', '=', model_name)], limit=1).field_ids.filtered(lambda field: field.value == 'date')
+            fieldnames = datefields.mapped('field_id').mapped('name')
+            docsource_data.update(dict(zip(fieldnames, [fields.Date.today(self)] * len(fieldnames))))
+
+            # if there was only one remaining document to generate
+            # the subscription is over and we mark it as being done
+            if subscription.cron_id.numbercall == 1:
+                subscription.write({'state': 'done'})
+            else:
+                subscription.write({'state': 'running'})
+            self.env[model_name].browse(subscription.doc_source.id).copy(docsource_data)
+            history_model.create({'subscription_id': subscription.id, 'history_date': fields.Datetime.now(), 'document_id': '%s,%s' % (subscription.doc_source._name, subscription.doc_source.id)})
```

Comment:
You are already having [model_name](https://github.com/odoo/odoo/pull/7164/files#diff-307bfd9c2e8ea3b2faa06c2060631021R80) variable. You can use it instead of subscription.doc_source._name.